### PR TITLE
feat: unified interface for da worker

### PR DESCRIPTION
**Pull request by drspacemn**
PR is based on unifying the 3 DA Implementations here:
[Avail](https://github.com/Leouarz/madara/pull/1/files)
[Celestia](https://github.com/keep-starknet-strange/madara/pull/967/files)
[Ethereum](https://github.com/keep-starknet-strange/madara/tree/pair-programming/exploration-team)

The primary interface draws logic from @Leouarz interface [here](https://github.com/keep-starknet-strange/madara/pull/1020)
(some methods internalized to the config initialization)
```rust
#[async_trait]
pub trait DaClient: Send + Sync {
    fn get_mode(&self) -> DaMode;
    async fn last_published_state(&self) -> Result<I256>;
    async fn publish_state_diff(&self, state_diff: Vec<U256>) -> Result<()>;
}
```
DaClient implementers are meant to be built from config.json files located at <madara_path>/da-config.json, but are also expected to have sensible defaults.

```rust
        let da_client: Box<dyn DaClient + Send + Sync> = match da_layer {
            DaLayer::Celestia => {
                let celestia_conf = CelestiaConfig::try_from_file(&da_path)?;
                Box::new(
                    CelestiaClient::try_from_config(celestia_conf).map_err(|e| ServiceError::Other(e.to_string()))?,
                )
            }
            DaLayer::Ethereum => {
                let ethereum_conf = EthereumConfig::try_from_file(&da_path)?;
                Box::new(EthereumClient::try_from_config(ethereum_conf)?)
            }
            DaLayer::Avail => {
                let avail_conf = AvailConfig::try_from_file(&da_path)?;
                Box::new(AvailClient::try_from_config(avail_conf).map_err(|e| ServiceError::Other(e.to_string()))?)
            }
        };
 ```

The Enum DaMode defines the potential data availability modes for Madara. As proving functionality is not currently functional, all modes default to Validium:

```rust
#[derive(Debug, Copy, Clone, PartialEq, Deserialize, Default)]
pub enum DaMode {
    #[serde(rename = "validity")]
    Validity, 
    #[serde(rename = "volition")]
    Volition,
    #[serde(rename = "validium")]
    #[default]
    Validium,
}
```

## Summary of Changes

| Component | Change                                                                                                                                       |
| --------- | -------------------------------------------------------------------------------------------------------------------------------------------- |
| CLI       | New optional flags in `ExtendedRunCommand`: `da-layer` [ethereum, celestia,avail] |
| Service   | state_diff notifications, push `state_diff` to the selected DA and verify that the `state_diff` output is included in the correct place. |

## Configuration file
When launching the node da config file is expected at `<madara-path>/da-config.json`

An example of running a node w/ DA.

```
RUST_LOG=info cargo run --release -- --da-layer=ethereum --dev
```

### Launch with Avail

***da-config.json***
```
{
   "ws_provider": "ws://127.0.0.1:9945",
   "app_id": 0,
   "validate_codegen": true,
   "seed": "//Alice"
}
```

You can find instruction to launch the [Avail node](https://github.com/availproject/avail). TLDR:
```
#  testnet ws wss://kate.avail.tools/ws
git clone https://github.com/availproject/avail
cd avail
cargo run --release -- --dev --tmp --rpc-port 9934 --ws-port 9945 --port 30334
```

### Launch with Celestia

***da-config.json***
```
{
   "http_provider": "http://127.0.0.1:26658",
   "ws_provider": "ws://127.0.0.1:26658",
   "nid": "Madara",
   "auth_token": "<see below>"
}
```
#### Set up and run a Celestia light node

Requires the installation of the celestia CLI tool (available [here](https://docs.celestia.org/developers/node-tutorial/)). Once you have that set up, run the following:

```
celestia light start --core.ip consensus-full-arabica-9.celestia-arabica.com --gateway.deprecated-endpoints --gateway --gateway.addr 127.0.0.1 --gateway.port 26659 --p2p.network arabica-9
```

Next, we need to generate a JWT to access our light node:

```
celestia light auth admin --p2p.network arabica-9
```

### Launch with Ethereum

***da-config.json***
```
{
   "http_provider": "http://127.0.0.1:8545",
   "core_contracts": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
   "sequencer_key": "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
   "chain_id": 31337
}
```

#### Start Anvil and Deploy altered Core Contracts

```
git clone --recurse-submodules https://github.com/keep-starknet-strange/zaun.git target/zaun 2> /dev/null
./target/zaun/scripts/sn-base-dev.sh target target/zaun 2> /dev/null
```

Integration Testing to be done via `scripts/da_devnet.sh`.

### CODE REVIEW

File: crates/client/data-availability/src/ethereum/config.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        serde_json::from_reader(file).map_err(|e| format!("error parsing da config: {e}"))
+    }
+}
+
+fn default_http() -> String {
+    format!("http://{DEFAULT_ETHEREUM_NODE}")
+}
+
+fn default_core_contracts() -> String {
+    DEFUALT_STARKNET_CORE_CONTRACTS.to_string()
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/config.rs:39]: typo DEFUALT

File: crates/client/data-availability/src/ethereum/config.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+use std::fs::File;
+use std::path::PathBuf;
+
+use serde::Deserialize;
+
+use crate::DaMode;
+
+pub const DEFAULT_ETHEREUM_NODE: &str = "127.0.0.1:8545";
+pub const DEFAULT_SEQUENCER_KEY: &str = "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/config.rs:9]: Where does this value come from?

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        let core_contracts = STARKNET::new(self.cc_address, self.signer.clone());
+
+        let fmt_tx = core_contracts.update_state(state_diff, U256::default(), U256::default());
+        let tx = fmt_tx
+            .send()
+            .await
+            .map_err(|e| anyhow::anyhow!("ethereum send update err: {e}"))?
+            .await
+            .map_err(|e| anyhow::anyhow!("ethereum poll update err: {e}"))?;
+
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:44]: Why don't we verify it was included?

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        contract.state_block_number().call().await.map_err(|e| anyhow::anyhow!("ethereum contract err: {e}"))
+    }
+
+    fn get_mode(&self) -> DaMode {
+        self.mode
+    }
+}
+
+impl EthereumClient {
+    pub fn try_from_config(conf: config::EthereumConfig) -> Result<Self, String> {
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:67]: Use TryFrom

File: crates/client/data-availability/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    pub async fn prove_current_block(da_mode: DaMode, client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
+        let mut storage_event_st = client.storage_changes_notification_stream(None, None).unwrap();
+
+        while let Some(storage_event) = storage_event_st.next().await {
+            // Locate and encode the storage change
+            let mut _deployed_contracts: Vec<String> = Vec::new();
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:60]: If it's never used, can we remove it?

File: crates/client/data-availability/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+impl<B, C> DataAvailabilityWorker<B, C>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    pub async fn prove_current_block(da_mode: DaMode, client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
+        let mut storage_event_st = client.storage_changes_notification_stream(None, None).unwrap();
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:56]: Why is it safe to unwrap?

File: crates/client/data-availability/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                log::error!("db err: {db_err}");
+            };
+
+            match da_mode {
+                DaMode::Validity => {
+                    // Submit the StarkNet OS PIE
+                    if let Ok(job_resp) = sharp::submit_pie("test") {
+                        log::info!("Job Submitted: {}", job_resp.cairo_job_key);
+                        // Store the cairo job key
+                        let _res = madara_backend.da().update_cairo_job(&storage_event.block, job_resp.cairo_job_key);
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:106]: Error is not handled? Not even a log

File: crates/client/data-availability/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+            // Store the DA output from the SN OS
+            if let Err(db_err) = madara_backend.da().store_state_diff(&storage_event.block, state_diff) {
+                log::error!("db err: {db_err}");
+            };
+
+            match da_mode {
+                DaMode::Validity => {
+                    // Submit the StarkNet OS PIE
+                    if let Ok(job_resp) = sharp::submit_pie("test") {
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:103]: Is it just a placeholder until further development?
Or should it be the app-chain responsibility to edit this code?

File: crates/client/data-availability/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    }
+}
+
+impl<B, C> DataAvailabilityWorker<B, C>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    // pub async fn update_state(client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:123]: Forgotten commented code

File: crates/client/data-availability/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    pub async fn update_state(
+        da_client: Box<dyn DaClient + Send + Sync>,
+        client: Arc<C>,
+        madara_backend: Arc<mc_db::Backend<B>>,
+    ) {
+        let mut notification_st = client.import_notification_stream();
+
+        while let Some(notification) = notification_st.next().await {
+            // Query last written state
+            let _last_published_state = match da_client.last_published_state().await {
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:133]: Why is the return from this query never used? Is the query still necessary?

Comment (tdelabro) [reply crates/client/data-availability/src/lib.rs:133]: Or is it only needed for `Validity` mode?

File: crates/client/data-availability/src/sharp/mod.rs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    let data = serde_json::json!({ "action": "add_job", "request": data });
+    let _payload: serde_json::Value = serde_json::from_value(data).unwrap();
+
+    // CAREFUL NOT TO OVERWHELM SHARP DUE TO SHORT BLOCK TIMES
+    // let resp = reqwest::blocking::Client::new().post(LAMBDA_URL).json(&payload).send().unwrap();
+
+    // match resp.status() {
+    //     reqwest::StatusCode::OK => Ok(resp.json::<CairoJobResponse>().unwrap()),
+    //     _ => Err(String::from("could not submit pie")),
+    // }
```
Comment (tdelabro) [inline crates/client/data-availability/src/sharp/mod.rs:55]: Should we uncomment this or not?

File: crates/client/data-availability/src/sharp/mod.rs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct CairoStatusResponse {
+    pub status: Option<String>,
+    #[serde(rename = "validation_done")]
+    pub validation_done: Option<bool>,
+    pub version: Option<u64>,
+}
+
+// Fetch Cairo Job Status from SHARP
+pub fn _get_status(job_key: &str) -> Result<CairoStatusResponse> {
```
Comment (tdelabro) [inline crates/client/data-availability/src/sharp/mod.rs:70]: Why is it never used nor exported?

File: crates/client/db/src/da_db.rs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+use crate::DbHash;
+
+// The fact db stores DA facts that need to be written to L1
+pub struct DaDb<B: BlockT> {
+    pub(crate) db: Arc<dyn Database<DbHash>>,
+    pub(crate) _marker: PhantomData<B>,
+}
+
+// TODO: business logic for last proven and purge
```
Comment (tdelabro) [inline crates/client/db/src/da_db.rs:19]: Has it been done?

File: crates/node/src/command.rs [hunk 14]
```diff
@@ -265,16 +268,28 @@ pub fn run() -> sc_cli::Result<()> {
...
-                cli.run.run_cmd.shared_params.chain = Some(madara_path + "/chain-specs/testnet-sharingan-raw.json");
+                cli.run.run_cmd.shared_params.chain =
+                    Some(madara_path.clone() + "/chain-specs/testnet-sharingan-raw.json");
 
                 // This should go apply to all testnets when applying a match pattern
                 cli.run.run_cmd.rpc_external = true;
                 cli.run.run_cmd.rpc_methods = RpcMethods::Unsafe;
             }
 
+            let mut da_config: Option<(DaLayer, PathBuf)> = None;
```
Comment (tdelabro) [inline crates/node/src/command.rs:279]: `let da_config = if let Some(da_layer) = cli.run.da_layer {} else { None };`
Or `match` if you find it more readable

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Review by EvolveArt — COMMENTED** [review]
A few questions on top of my mind :

1. What happens if the madara nodes see that the DA was not updated correctly ?
2. What does it mean to be in Validity `DaMode` for a client like `CelestiaClient` ? Does it just mean we also publish state updates on L1 ?

File: crates/client/data-availability/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+#[derive(Debug, Copy, Clone, PartialEq, Deserialize, Default)]
+pub enum DaMode {
+    #[serde(rename = "validity")]
+    Validity,
+    #[serde(rename = "volition")]
+    Volition,
+    #[serde(rename = "validium")]
+    #[default]
+    Validium,
+}
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:40]: Can we have a lil' bit of doc on what those modes imply?

**Comment by tdelabro** [comment]
@EvolveArt 
> What does it mean to be in Validity DaMode for a client like CelestiaClient? Does it just mean we also publish state updates on L1 ?

I think it means we publish the state, but also prove the state update computation

File: crates/client/data-availability/src/avail/mod.rs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        let bytes = BoundedVec(bytes);
+
+        let submitted_block_hash = self.publish_data(&bytes).await?;
+
+        self.verify_bytes_inclusion(submitted_block_hash, &bytes).await?;
+        Ok(())
+    }
+
+    async fn last_published_state(&self) -> Result<I256> {
+        Ok(I256::from(1))
```
[the remark this answers is not in the document]
Comment (drspacemn) [reply crates/client/data-availability/src/avail/mod.rs:43]: added as comment

File: crates/client/data-availability/src/lib.rs (hunk 10 above)
Comment (drspacemn) [reply crates/client/data-availability/src/lib.rs:134]: `// TODO: this value will be used to ensure the correct state diff is being written in Validity mode`

added as comment

File: crates/client/data-availability/src/avail/mod.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (drspacemn) [reply crates/client/data-availability/src/avail/mod.rs:43]: `    // state diff can be published w/o verification of last state for the time being
    // may change in subsequent DaMode implementations`
    
    Added as comment

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    http_provider: Provider<Http>,
+    signer: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    cc_address: Address,
+    mode: DaMode,
+}
+
+#[async_trait]
+impl DaClient for EthereumClient {
+    async fn publish_state_diff(&self, state_diff: Vec<U256>) -> Result<()> {
+        println!("State diff: {:?}", state_diff);
```
[the remark this answers is not in the document]
Comment (drspacemn) [reply crates/client/data-availability/src/ethereum/mod.rs:26]: indeed

File: crates/client/data-availability/src/ethereum/config.rs (hunk 2 above)
Comment (drspacemn) [reply crates/client/data-availability/src/ethereum/config.rs:9]: // default key derived from starting anvil as follows:
// anvil -b 5 --config-out $BUILD_DIR/anvil.json
// PRE_PRIVATE=$(jq -r '.private_keys[0]' $BUILD_DIR/anvil.json)

added as comment

File: crates/client/data-availability/src/lib.rs (hunk 6 above)
Comment (drspacemn) [reply crates/client/data-availability/src/lib.rs:56]: Can update to log this storage change notification failure, but we initialize the node with the intent to prove thee current block and then subsequently can't read from the storage change stream we are operating in a bad state.

Would be on the operator to check the logs and see that no proofs are being generated.

Comment (drspacemn) [reply crates/client/data-availability/src/lib.rs:56]: Added an expect for the time being

File: crates/client/data-availability/src/lib.rs (hunk 8 above)
Comment (drspacemn) [reply crates/client/data-availability/src/lib.rs:103]: // TODO: Validity Impl
                    // run the Starknet OS with the Cairo VM
                    // extract the PIE from the Cairo VM run
                    // pass the PIE to `submit_pie` and zip/base64 internal
  
  added comment

File: crates/client/data-availability/src/sharp/mod.rs (hunk 11 above)
Comment (drspacemn) [reply crates/client/data-availability/src/sharp/mod.rs:55]: I think we should leave this commented with a TODO to uncomment for Validity impl.

added comment

File: crates/client/data-availability/src/sharp/mod.rs (hunk 12 above)
Comment (drspacemn) [reply crates/client/data-availability/src/sharp/mod.rs:70]: // TODO: function will be needed in Validity DaMode impl

added comment

File: crates/client/db/src/da_db.rs (hunk 13 above)
Comment (drspacemn) [reply crates/client/db/src/da_db.rs:19]: no, updated comment

File: crates/client/data-availability/src/lib.rs (hunk 15 above)
Comment (drspacemn) [reply crates/client/data-availability/src/lib.rs:40]: doc added

**Comment by drspacemn** [comment]
> A few questions on top of my mind :
> 
> 1. What happens if the madara nodes see that the DA was not updated correctly ?
> 2. What does it mean to be in Validity `DaMode` for a client like `CelestiaClient` ? Does it just mean we also publish state updates on L1 ?

1. This will be the tradeoffs of the DaModes. If madara node is `Validity` the Core Contracts (or other solutions see 2) will reject that update. If its Validium, yeah that's the rub, but operators can determine their own consensus around data that is now available on say Celestia

2. Yet to be fully defined, but there are solutions like:
 - [celestia's quantum gravity bridge](https://github.com/celestiaorg/quantum-gravity-bridge)
 - [avail's attestation bridge](https://availproject.github.io/using-avail/validiums#verify-data-availability-on-ethereum)
 
 that can allow us to implement a version of Validity

**Review by tdelabro — APPROVED** [review]

**Comment by tdelabro** [comment]
@drspacemn just fix the CI and rebase and we can merge

**Comment by Leouarz** [comment]
> > A few questions on top of my mind :
> > 
> > 1. What happens if the madara nodes see that the DA was not updated correctly ?
> > 2. What does it mean to be in Validity `DaMode` for a client like `CelestiaClient` ? Does it just mean we also publish state updates on L1 ?
> 
> 1. This will be the tradeoffs of the DaModes. If madara node is `Validity` the Core Contracts (or other solutions see 2) will reject that update. If its Validium, yeah that's the rub, but operators can determine their own consensus around data that is now available on say Celestia
> 2. Yet to be fully defined, but there are solutions like:
> 
> * [celestia's quantum gravity bridge](https://github.com/celestiaorg/quantum-gravity-bridge)
> * [avail's attestation bridge](https://availproject.github.io/using-avail/validiums#verify-data-availability-on-ethereum)
> 
> that can allow us to implement a version of Validity

Here's the correct link for Avail btw : https://availproject.github.io/api/use-case-validiums/#verify-data-availability-on-ethereum

**Comment by drspacemn** [comment]
> > > A few questions on top of my mind :
> > > 
> > > 1. What happens if the madara nodes see that the DA was not updated correctly ?
> > > 2. What does it mean to be in Validity `DaMode` for a client like `CelestiaClient` ? Does it just mean we also publish state updates on L1 ?
> > 
> > 
> > 
> > 1. This will be the tradeoffs of the DaModes. If madara node is `Validity` the Core Contracts (or other solutions see 2) will reject that update. If its Validium, yeah that's the rub, but operators can determine their own consensus around data that is now available on say Celestia
> > 2. Yet to be fully defined, but there are solutions like:
> > 
> > 
> > * [celestia's quantum gravity bridge](https://github.com/celestiaorg/quantum-gravity-bridge)
> > * [avail's attestation bridge](https://availproject.github.io/using-avail/validiums#verify-data-availability-on-ethereum)
> > 
> > that can allow us to implement a version of Validity
> 
> @drspacemn I still find a few things confusing here regarding SHARP/DA coupling
> 
> 1. Imo, SHARP and DA (or DP as now people want to say) should be decoupled and not necessarily done as part of same thread
> 2. In the current SHARP integration we directly submit the cairo_pie to SHARP, I know it's mocked for now but we need to run the starknet OS program somewhere in the middle right ?
> 3. What happens if the app-chain wants to switch of DA Layer, is there anyway to retro-actively publish DA on the new layer ? Would potentially also apply to state migration

1.  SHARP and DA are decoupled in this PR. There is a task spawned to handle the state_update and a task spawned to handle the proving. both will handle their business logic via the mode they have been initialized in. 

2. Yes the starknet OS program will need to be run in the `prove_current_block` task and only if its applicable to the `match da_mode {` (i.e. not for Valium)

3. That is a good question probably out of scope for the current work done here.

**Comment by EvolveArt** [comment]
Sounds good, would be great to document all of these questions and features to work on, probably the best place is a GH discussion, wdyt @drspacemn ?

Otherwise let's merge this PR which is a great starting point to on-board new contributors and great work on this again!

**Review by tdelabro — APPROVED** [review]

**Pull request merged**